### DISCUSSION TITLE
How to export C types in CFFI

### DISCUSSION THREAD

**Discussion by michaelserrx** [cffi/cffi#395] [Q&A]
Hi all, first of all sorry if the answer of this question is obvious but I'm fairly new to both CFFI and Common Lisp programming in general. Basically I imported a type definition from a C library with `(defctype window-ptr :pointer)`.
The problem is, when I try to use it as a type as the `:type` keyword parameter in a `defstruct`, SBCL complains with:
```
#<THREAD tid=8963 "main thread" RUNNING {1003E801D3}>:
  unknown type specifier: (NCURSES:WINDOW-PTR)
```
Do you know if and how I can use this C type as a type specifier for `defstruct`? Thanks in advance.

**Comment by luismbo** [comment]
Indeed, defctype does not define a Lisp type. What were you hoping to achieve with your defstruct? Do you want defcstruct instead?

**Reply by michaelserrx** [reply]
Basically I am writing bindings for linbcurses with cffi, and I was creating a Lisp struct `ui` that containd a couple of windows created with `newwin`:
```
(defstruct ui                                               
  (main-win (ncurses:newwin                                 
              (- ncurses:*lines* +buffer-win-height+)       
              ncurses:*cols*                                
              0 0))                  
  (input-win (ncurses:newwin                                
               +buffer-win-height+                          
               ncurses:*cols*                               
               (- ncurses::*lines* +buffer-win-height+) 0)))                
```
where `ncurses` is the package containing the CFFI bindings. I wanted to enforce the type of the windows as `window-ptr` with `:type` attribute in Lisp struct but don't know how to do it since CFFI does not define Lisp types.

**Reply by luismbo** [reply]
The best you can do is declare it as a `cffi:foreign-pointer`. CFFI's pointers are untyped. It'd be nice to build an API like SBCL/CMUCL's (sb-)alien or Lispworks FLI on top of it!
